src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java

```diff
@@ -37,6 +42,85 @@ static String getFilenameForKey(String key) {
         return localFilename;
     }
 
+    /** Returns a file object for the given cache key. */
+    public static File getFileForKey(String key, FileSupplier rootDirectorySupplier) {
+        return new File(rootDirectorySupplier.get(), getFilenameForKey(key));
+    }
+
+    /**
+     * Prunes the .
+     *
+     * @param totalSize The total size of the cache.
+     * @param maxCacheSizeInBytes Maximum size of the cache.
+     * @param entries Map of the entries in the cache.
+     * @param rootDirectorySupplier The supplier for the root directory to use for the cache.
+     * @return A long to update the totalSize.
+     */
+    static long pruneIfNeeded(
+            long totalSize,
+            int maxCacheSizeInBytes,
+            Map<String, CacheHeader> entries,
+            FileSupplier rootDirectorySupplier) {
+        if (totalSize < maxCacheSizeInBytes) {
+            return totalSize;
+        }
+        if (VolleyLog.DEBUG) {
+            VolleyLog.v("Pruning old cache entries.");
+        }
+
+        long before = totalSize;
+        int prunedFiles = 0;
+        long startTime = SystemClock.elapsedRealtime();
+
+        Iterator<Map.Entry<String, CacheHeader>> iterator = entries.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, CacheHeader> entry = iterator.next();
+            CacheHeader e = entry.getValue();
+            boolean deleted = getFileForKey(e.key, rootDirectorySupplier).delete();
+            if (deleted) {
+                totalSize -= e.size;
+            } else {
+                VolleyLog.d(
+                        "Could not delete cache entry for key=%s, filename=%s",
+                        e.key, DiskBasedCacheUtility.getFilenameForKey(e.key));
+            }
+            iterator.remove();
+            prunedFiles++;
+
+            if (totalSize < maxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
+                break;
+            }
+        }
+
+        if (VolleyLog.DEBUG) {
+            VolleyLog.v(
+                    "pruned %d files, %d bytes, %d ms",
+                    prunedFiles, (totalSize - before), SystemClock.elapsedRealtime() - startTime);
+        }
+        return totalSize;
+    }
+
+    /**
+     * Puts the entry with the specified key into the cache.
+     *
+     * @param key The key to identify the entry by.
+     * @param entry The entry to cache.
+     * @param totalSize The total size of the cache.
+     * @param entries Map of the entries in the cache.
+     * @return A long to update the total size.
+     */
+    static long putEntry(
+            String key, CacheHeader entry, long totalSize, Map<String, CacheHeader> entries) {
```

Comment:
I understand why you're passing `entries` here, now that it's static, but keep in mind that it's generally not a good practice to modify the arguments of a method, as they're interpreted as an input to the function and not something that will be modified. The user will have to check the doc (which btw should reflect that `entries` will be updated`) to fully understand what the method does (and while it's always a good idea to do so, you can't guarantee future developers will do so, and it's always better to go with self-explicit methods).

I could be convinced that it's fine to keep this as is because it's internal, but we would need a more descriptive documentation. Alternatively, you could keep the method to compute the new total size, and do the `entries.put(key, entry)` at the caller site instead. I would personally prefer this approach.

Reply:
Would this be true for pruneIfNeeded as well, since it also changes the entries map? And there will probably be a couple more methods like this (removeEntry, etc.) eventually. Since this would duplicate a lot of code, do you still think that this would be the better route?

Also, we could have the methods return both the entries map and the updated totalSize, then set it in the caller. What're your thoughts?

Reply:
I thought about returning both, but I am not sure it would make the code more readable, so we can stick to this if you think it's better this way.